gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java

```diff
@@ -69,108 +70,93 @@ public RetryResult beforeEval(Exception exception) {
   static final ExceptionHandler EXCEPTION_HANDLER = ExceptionHandler.builder()
       .abortOn(RuntimeException.class).interceptor(EXCEPTION_HANDLER_INTERCEPTOR).build();
 
-  private abstract static class BasePageFetcher<T> implements PageImpl.NextPageFetcher<T> {
+  private static class DatasetPageFetcher extends BasePageFetcher<DatasetInfo, BigQueryOptions> {
 
-    private static final long serialVersionUID = -338124488600215401L;
-
-    protected final Map<BigQueryRpc.Option, ?> requestOptions;
-    protected final BigQueryOptions serviceOptions;
-
-    BasePageFetcher(BigQueryOptions serviceOptions, String cursor,
-        Map<BigQueryRpc.Option, ?> optionMap) {
-      this.serviceOptions = serviceOptions;
-      ImmutableMap.Builder<BigQueryRpc.Option, Object> builder = ImmutableMap.builder();
-      if (cursor != null) {
-        builder.put(BigQueryRpc.Option.PAGE_TOKEN, cursor);
-      }
-      for (Map.Entry<BigQueryRpc.Option, ?> option : optionMap.entrySet()) {
-        if (option.getKey() != BigQueryRpc.Option.PAGE_TOKEN) {
-          builder.put(option.getKey(), option.getValue());
-        }
-      }
-      this.requestOptions = builder.build();
-    }
-  }
-
-  private static class DatasetPageFetcher extends BasePageFetcher<DatasetInfo> {
-
-    private static final long serialVersionUID = 3030824397616608646L;
+    private static final long serialVersionUID = 8049465230453759547L;
 
     DatasetPageFetcher(BigQueryOptions serviceOptions, String cursor,
         Map<BigQueryRpc.Option, ?> optionMap) {
-      super(serviceOptions, cursor, optionMap);
+      super(serviceOptions, BigQueryRpc.Option.PAGE_TOKEN, cursor, optionMap);
     }
 
+    @SuppressWarnings("unchecked")
```

Comment:
I wonder if instead of implementing the common code in the base class `BasePageFetcher` (and suppressing its rawtype warning) and then needing to cast and suppress warning in each subclass we would not be better of with putting the common logic in a static method.
```java
  /**
   * Utility method to construct the next options map.
   *
   * @param <T> the value type that the page holds. Instances of {@code T} should be
   *     {@code Serializable}
   * @param pageTokenOption the key for the
   * @param cursor the cursor for the next page
   * @param optionMap the previous options map
   */
  public static <T> Map<T, Object> nextRequestOptions(T pageTokenOption, String cursor,
      Map<T, ?> optionMap) {
    ImmutableMap.Builder<T, Object> builder = ImmutableMap.builder();
    if (cursor != null) {
      builder.put(pageTokenOption, cursor);
    }
    for (Map.Entry<T, ?> option : optionMap.entrySet()) {
      if (!Objects.equals(option.getKey(), pageTokenOption)) {
        builder.put(option.getKey(), option.getValue());
      }
    }
    return builder.build();
  }
```

and In the implementations of `NextPageFetcher` constructors we would instead call that static method to get the next requestOptions (and keep a reference to it as well as to the specific service options which was not needed by the common code).

e.g.
```java
  private static class ProjectPageFetcher implements NextPageFetcher<ProjectInfo> {

    private static final long serialVersionUID = -533306655445189098L;
    private final Map<ResourceManagerRpc.Option, ?> requestOptions;
    private final ResourceManagerOptions serviceOptions;

    ProjectPageFetcher(ResourceManagerOptions serviceOptions, String cursor,
        Map<ResourceManagerRpc.Option, ?> optionMap) {
      this.requestOptions =
          PageImpl.nextRequestOptions(ResourceManagerRpc.Option.PAGE_TOKEN, cursor, optionMap);
      this.serviceOptions = serviceOptions;
    }

    @Override
    public Page<ProjectInfo> nextPage() {
      return listProjects(serviceOptions, requestOptions);
    }
  }
```